File: 1054-distant-barcodes/1054-distant-barcodes.cpp
class Solution {
public:
    vector<int> rearrangeBarcodes(vector<int>& barcodes) {

        unordered_map<int, int> freq;

        // Count frequencies
        for (int barcode : barcodes) {
            freq[barcode]++;
        }

        // Convert to {barcode, frequency}
        vector<pair<int, int>> items;

        for (auto& [barcode, count] : freq) {
            items.push_back({barcode, count});
        }

        // Sort by frequency descending
        sort(items.begin(), items.end(),
             [](const pair<int, int>& a,
                const pair<int, int>& b) {
                 return a.second > b.second;
             });

        vector<int> result(barcodes.size());

        int index = 0;

        for (auto& [barcode, count] : items) {

            while (count > 0) {

                result[index] = barcode;

                count--;

                // Move to next even position
                index += 2;

                // If even positions are exhausted,
                // start from the first odd position.
                if (index >= result.size()) {
                    index = 1;
                }
            }
        }

        return result;
    }
};

File: 1054-distant-barcodes/1054-distant-barcodes.java
class Solution {
    public int[] rearrangeBarcodes(int[] barcodes) {

        // Step 1: Count frequency
        Map<Integer, Integer> freq = new HashMap<>();

        for (int barcode : barcodes) {
            freq.put(
                barcode,
                freq.getOrDefault(barcode, 0) + 1
            );
        }

        // Step 2: Sort barcodes by frequency descending
        List<int[]> items = new ArrayList<>();

        for (Map.Entry<Integer, Integer> entry : freq.entrySet()) {
            items.add(new int[]{
                entry.getKey(),
                entry.getValue()
            });
        }

        items.sort((a, b) -> Integer.compare(b[1], a[1]));

        // Step 3: Fill answer
        int[] result = new int[barcodes.length];

        int index = 0;

        for (int[] item : items) {

            int barcode = item[0];
            int count = item[1];

            while (count > 0) {

                result[index] = barcode;

                count--;

                // Move by 2 so identical values
                // are never adjacent.
                index += 2;

                // Even positions exhausted.
                // Start filling odd positions.
                if (index >= result.length) {
                    index = 1;
                }
            }
        }

        return result;
    }
}